src/main.cpp

```diff
@@ -2501,21 +2530,90 @@ void looppid() {
 }
 
 void loopLED() {
-    if(FEATURE_STATUS_LED) {
-        if ((machineState == kPidNormal && (fabs(temperature  - setpoint) < 0.3)) || (temperature > 115 && fabs(temperature - setpoint) < 5)) {
-            statusLed->turnOn();
-        }
-        else {
-            statusLed->turnOff();
+    if (LEDupdate.check()) {
+
+        if(FEATURE_STATUS_LED) {
+            if ((machineState == kPidNormal && (fabs(temperature  - setpoint) < 0.3)) || (temperature > 115 && fabs(temperature - setpoint) < 5)) {
+                statusLed->turnOn();
+            }
+            else {
+                statusLed->turnOff();
+            }
         }
-    }
 
-    if (FEATURE_BREW_LED) {
-        if (machineState == kBrew) {
-            brewLed->turnOn();
+        if (FEATURE_BREW_LED) {
+            if (machineState == kBrew) {
+                brewLed->turnOn();
+            }
+            else {
+                brewLed->turnOff();
+            }
         }
-        else {
-            brewLed->turnOff();
+
+        if (FEATURE_CONTEXT_LED) {
```

Comment:
I suspect you're trying to get around having to write if (FEATURE_CONTEXT_LED) in every case-statement but duplicating the whole state machine seems infinitely worse.

Reply:
I started in the state machine, but its quite a lot of code duplication, as e.g. 
                case kBrew:
                case kShotTimerAfterBrew:
                case kBrewDetectionTrailing:
                case kBackflush:  
all have the same behaviour and the code below these the four cases would have to be repeated in four different places (in each "case" of the main state machine)

On top for each case there should be:
- The "if FEATURE.." check
- The "LEDupdate.check()"

Readability is also better, its easy to see what the behaviour is, and all LED code is in the loopLED() call.

Also it would not fit into the main state machine perfectly, as i want to implement a "cup illumination" for a longer time with smooth fade out. That wouldnt easily be possible in one location of the state machine, but rather a check before the loopLED switch case.

Reply:
Tried to reduce the switch-case to the maximum, still it seems cleaner to keep it in the same location (would need check if an error state is set to not overwrite two different colors with each other each loop().